File: Leetcode100/leetcode100/src/main/java/com/jiang/leetcode100/letter_combinations_of_a_phone_number/LetterCombinationsofaPhoneNumber2.java
package main.java.com.jiang.leetcode100.letter_combinations_of_a_phone_number;

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

class Solution2 {

    String[] mapping = new String[]{
        "abc",
        "def",
        "ghi",
        "jkl",
        "mno",
        "pqrs",
        "tuv",
        "wxyz"
    };

    public List<String> letterCombinations(String digits) {

        List<String> result = new ArrayList<>();
        String combinations = "";
        dfs(digits, 0, combinations, result);
        return result;
    }

    public void dfs(String digits, int index, String combinations,
        List<String> result) {
        
        if(index == digits.length()) {
            if(!combinations.isEmpty()) result.add(combinations);
            return;
        }

        String cur_chars = mapping[digits.charAt(index) - '2'];
        for(int i = 0; i < cur_chars.length(); i++) {
            dfs(digits, index + 1, combinations + cur_chars.charAt(i), result);
        }

    }
}

public class LetterCombinationsofaPhoneNumber2 {
    
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String digits = scanner.nextLine();
        scanner.close();

        List<String> result = new Solution2().letterCombinations(digits);
        for(String s : result) {
            System.out.println(s);
        }
    }

}

File: Leetcode100/leetcode100_python/LetterCombinationsofaPhoneNumber2.py
from typing import List

class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        def dfs(index, combinations):
            if index == len(digits):
                result.append(combinations)
                return
            for ch in mapping[digits[index]]:
                dfs(index + 1, combinations + ch)
        if not digits:
            return []
        mapping = {
            '2': "abc",
            '3': "def",
            '4': "ghi",
            '5': "jkl",
            '6': "mno",
            '7': "pqrs",
            '8': "tuv",
            '9': "wxyz"}
        result = []
        dfs(0, "")
        return result

if __name__ == "__main__":
    digits = input()
    solution = Solution()
    print(solution.letterCombinations(digits))
        